components/org.wso2.carbon.identity.oauth/src/test/java/org/wso2/carbon/identity/oauth2/OAuth2ServiceTest.java

```diff
@@ -176,26 +188,78 @@ public void testAuthorizeWithException() throws IdentityOAuth2Exception {
+        assertNotNull(oAuth2ClientValidationResponseDTO);
+    }
+
+    @Test
+    public void testValidateClientInfoWithEmptyGrantTypes() throws Exception {
+
+        getOAuthAppDO(null, "dummyCallbackUrl", "dummyTenantDomain", "dummyCallbackURI");
+        OAuth2ClientValidationResponseDTO oAuth2ClientValidationResponseDTO = oAuth2Service.
+                validateClientInfo(clientId, "dummyCallBackUrl");
+        assertEquals(oAuth2ClientValidationResponseDTO.getErrorCode(), "unauthorized_client");
+    }
+
+    private OAuthAppDO getOAuthAppDO(String grantType, String callbackUrl, String tenantDomain, String callbackURI)
+            throws Exception {
+
         whenNew(OAuthAppDAO.class).withNoArguments().thenReturn(oAuthAppDAO);
+        OAuthAppDO oAuthAppDO = new OAuthAppDO();
+        oAuthAppDO.setGrantTypes(grantType);
+        oAuthAppDO.setState("ACTIVE");
+        oAuthAppDO.setCallbackUrl(callbackUrl);
+        oAuthAppDO.setAppOwner(new AuthenticatedUser());
         when(oAuthAppDAO.getAppInformation(clientId)).thenReturn(oAuthAppDO);
-        when(oAuthAppDO.getGrantTypes()).thenReturn(grantType);
-        when(oAuthAppDO.getCallbackUrl()).thenReturn(callbackUrl);
-        when(oAuthAppDO.getUser()).thenReturn(authenticatedUser);
+        AppInfoCache.getInstance().addToCache(clientId, null);
         when(authenticatedUser.getTenantDomain()).thenReturn(tenantDomain);
         mockStatic(IdentityTenantUtil.class);
         when(IdentityTenantUtil.getTenantId(anyString())).thenReturn(1);
-        mockStatic(OAuthServerConfiguration.class);
-        when(OAuthServerConfiguration.getInstance()).thenReturn(oAuthServerConfiguration);
+        return oAuthAppDO;
+    }
+
+    @Test
+    public void testValidateClientInfoWithInvalidClientId() throws Exception {
+
+        whenNew(OAuthAppDAO.class).withNoArguments().thenReturn(oAuthAppDAO);
+        when(oAuthAppDAO.getAppInformation(null)).thenReturn(null);
         OAuth2ClientValidationResponseDTO oAuth2ClientValidationResponseDTO = oAuth2Service.
-                validateClientInfo(clientId, callbackURI);
+                validateClientInfo(null, "dummyCallbackUrI");
         assertNotNull(oAuth2ClientValidationResponseDTO);
+        assertEquals(oAuth2ClientValidationResponseDTO.getErrorCode(), "invalid_client");
+        assertFalse(oAuth2ClientValidationResponseDTO.isValidClient());
+    }
+
+    @DataProvider(name = "InvalidAppStatDataProvider")
+    public Object[][] invalidAppStateDataProvider() {
+
+        return new Object[][]{
+                {null},
+                {"dummyAppState"}
+        };
+    }
+
+    @Test(dataProvider = "InvalidAppStatDataProvider")
+    public void testValidateClientInfoWithInvalidAppState(String appState) throws Exception {
+
+        OAuthAppDO oAuthAppDO = getOAuthAppDO("dummyGrantType", "dummyCallbackUrl",
+                "dummyTenantDomain", "dummyCallbackURI");
+        oAuthAppDO.setState(appState);
```

Comment:
where do we use this `oAuthAppDO` object?

Reply:
The logic is written in getOAuthAppDO private method.
-> when(oAuthAppDAO.getAppInformation(clientId)).thenReturn(oAuthAppDO);